Python ClientWrapperTest fails with protobuf 3.14.0

This has apparently been a known issue for a while, but I've narrowed down the circumstances it can happen in while working on OpenLightingProject/website#14.  Warning - it's a bit involved.

When protobuf is installed from pip (i.e. `pip3 install protobuf`), ClientWrapperTest fails.  When protobuf is installed from the system package manager (i.e. `apt-get install python3-protobuf`), the test passes.  Checked with the latest commit in master (a95a6f7 currently) with Ubuntu 20.10 and Ubuntu 20.04.  I haven't properly checked Fedora yet (only a cursory test while writing docs), but I suspect it will have the same issue for reasons detailed below.

Some speculation:

My first thought was an issue between Python2 packages and Python3 packages.  These test systems only have Python3, so that's not it.

Further research reveals that GNOME distros (or, at least, Ubuntu and Fedora) come with `gnome-online-accounts`, which itself depends on `python3-protobuf`.  This means that GNOME users will likely have protobuf for python already installed on their systems.  If they install deps from pip (e.g. with `pip3 install protobuf numpy`), pip won't install another protobuf as it sees the system version already.  This also explains why the issue occurs on headless systems (i.e. Travis), as those systems won't have the preinstalled python protobuf and will install it from pip if you ask.

But what's the difference between the two sources?  My instinct is that something changed in newer Protobuf releases (I know this is a regular problem here).  The latest release on pip is [3.14.0](https://pypi.org/project/protobuf/#history).  Ubuntu 20.10 will get you [3.12.3](https://packages.ubuntu.com/groovy/python3-protobuf).  Ubuntu 20.04 provides [3.6.1](https://packages.ubuntu.com/focal/python3-protobuf).  Fedora 33 has [3.12.4](https://fedora.pkgs.org/33/fedora-x86_64/python3-protobuf-3.12.4-1.fc33.noarch.rpm.html).  Fedora 32 has [3.11.2](https://fedora.pkgs.org/32/fedora-x86_64/python3-protobuf-3.11.2-2.fc32.noarch.rpm.html).

To confirm this theory, I installed protobuf 3.12.4 from pip and ran tests, which all passed.  protobuf 3.13.0 also passes.  This leads me to conclude something about protobuf 3.14 is causing problems.  Caveat - I havn't tested any of this with Python2.

Test log:
```
==================================
   OLA 0.10.8: ./test-suite.log
==================================

# TOTAL: 95
# PASS:  94
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: python/ola/ClientWrapperTest.sh
=====================================

...F/usr/lib/python3.8/unittest/case.py:704: ResourceWarning: unclosed <socket.socket fd=4, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
  outcome.errors.clear()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
.
======================================================================
FAIL: testEventLoop (__main__.ClientWrapperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./python/ola/ClientWrapperTest.py", line 195, in testEventLoop
    self.assertTrue(results.d_called - self.start >=
AssertionError: False is not true

----------------------------------------------------------------------
Ran 5 tests in 0.039s

FAILED (failures=1)
FAIL python/ola/ClientWrapperTest.sh (exit status: 1)
```

I tweaked the failing test a bit to get better info:

```diff
diff --git a/python/ola/ClientWrapperTest.py b/python/ola/ClientWrapperTest.py
index cd1e5cf2d..a24041d15 100644
--- a/python/ola/ClientWrapperTest.py
+++ b/python/ola/ClientWrapperTest.py
@@ -186,13 +186,13 @@ class ClientWrapperTest(unittest.TestCase):
     self.assertIsNotNone(results.c_called)
     self.assertIsNotNone(results.d_called)
 
-    self.assertTrue(results.a_called - self.start <
+    self.assertLess(results.a_called - self.start,
                     datetime.timedelta(milliseconds=5))
-    self.assertTrue(results.b_called - self.start >=
+    self.assertGreaterEqual(results.b_called - self.start,
                     datetime.timedelta(milliseconds=5))
-    self.assertTrue(results.c_called - self.start >=
+    self.assertGreaterEqual(results.c_called - self.start,
                     datetime.timedelta(milliseconds=10))
-    self.assertTrue(results.d_called - self.start >=
+    self.assertGreaterEqual(results.d_called - self.start,
                     datetime.timedelta(milliseconds=15))
 
     sockets[0].close()
```
With the new test that says what the difference is, it fails with:

```
AssertionError: datetime.timedelta(microseconds=14612) not greater than or equal to datetime.timedelta(microseconds=15000)
```

However, I ran the same test a couple more times, without changing anything, and it passed.  Are we just dealing with system scheduler variances?

Glad you're investigating.  Though I don't know what protobuf could have to
do with either of those failures.

On Thu, Feb 18, 2021 at 1:09 PM Dan Keenan <notifications@github.com> wrote:

>
>
> FAIL: python/ola/ClientWrapperTest.sh
> =====================================
>
> ...F/usr/lib/python3.8/unittest/case.py:704: ResourceWarning: unclosed <socket.socket fd=4, family=AddressFamily.AF_UNIX, type=SocketKind.SOCK_STREAM, proto=0>
>   outcome.errors.clear()
> ResourceWarning: Enable tracemalloc to get the object allocation traceback
> ..
>
>
Glancing at the tests, the last test (testSend) doesn't have the
    sockets[0].close()
    sockets[1].close()
at the end.  That looks like my mistake.


>
> ======================================================================
> FAIL: testEventLoop (__main__.ClientWrapperTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "./python/ola/ClientWrapperTest.py", line 195, in testEventLoop
>     self.assertTrue(results.d_called - self.start >=
> AssertionError: False is not true
>
>
> Only way I see that happening would be a  strange rounding error, but I'd
be really surprised it's repeatable.  Curious what d_called and start are.
I considered putting in checks that were a bit less than the requested
delays (so 14ms here) but never saw it fail.

Bruce

(messages overlapped...)

I don't see how, as the check for when to call the task is explicit in the
code (it's not just a sleep):
    return self._run_at < now
but I wonder if there's some issue with timer precision underlying the
python implementation giving different results in different places.

I would just subtract 1 from each of the comparisons and leave it at that.

Bruce


On Thu, Feb 18, 2021 at 3:26 PM Dan Keenan <notifications@github.com> wrote:

> I tweaked the failing test a bit to get better info:
>
> diff --git a/python/ola/ClientWrapperTest.py b/python/ola/ClientWrapperTest.py
> index cd1e5cf2d..a24041d15 100644--- a/python/ola/ClientWrapperTest.py+++ b/python/ola/ClientWrapperTest.py@@ -186,13 +186,13 @@ class ClientWrapperTest(unittest.TestCase):
>      self.assertIsNotNone(results.c_called)
>      self.assertIsNotNone(results.d_called)
>  -    self.assertTrue(results.a_called - self.start <+    self.assertLess(results.a_called - self.start,
>                      datetime.timedelta(milliseconds=5))-    self.assertTrue(results.b_called - self.start >=+    self.assertGreaterEqual(results.b_called - self.start,
>                      datetime.timedelta(milliseconds=5))-    self.assertTrue(results.c_called - self.start >=+    self.assertGreaterEqual(results.c_called - self.start,
>                      datetime.timedelta(milliseconds=10))-    self.assertTrue(results.d_called - self.start >=+    self.assertGreaterEqual(results.d_called - self.start,
>                      datetime.timedelta(milliseconds=15))
>
>      sockets[0].close()
>
> With the new test that says what the difference is, it fails with:
>
> AssertionError: datetime.timedelta(microseconds=14612) not greater than or equal to datetime.timedelta(microseconds=15000)
>
> However, I ran the same test a couple more times, without changing
> anything, and it passed. Are we just dealing with system scheduler
> variances?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenLightingProject/ola/issues/1714#issuecomment-781613819>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADNHN5NWKT2RYBDWBEQSH5TS7VZXNANCNFSM4X2YUA5A>
> .
>

Oops, I thought I'd raised this elsewhere and tagged @brucelowekamp for some input but clearly not.

I hit the same thing on the Arm builds on Travis, but only confusingly at some point later the commits to get them merged worked fine, so I guess after some system change or similar, or package upgrade. I get (from https://travis-ci.org/github/OpenLightingProject/ola/jobs/755992758 which is a build of https://github.com/OpenLightingProject/ola/pull/1687):
```

FAIL: testEventLoop (__main__.ClientWrapperTest)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "../../python/ola/ClientWrapperTest.py", line 193, in testEventLoop

    datetime.timedelta(milliseconds=5))

AssertionError: datetime.timedelta(0, 0, 4923) not greater than or equal to datetime.timedelta(0, 0, 5000)
```

--- a/python/ola/ClientWrapperTest.py
+++ b/python/ola/ClientWrapperTest.py
@@ -191,7 +191,7 @@ class ClientWrapperTest(unittest.TestCase):
     # Check when the callbacks were called. Allow 500 microseconds of drift.
     # Called immediately
     a_diff = results.a_called - self.start
-    self.assertAlmostEqual(a_diff, datetime.timedelta(microseconds=0),
+    self.assertAlmostEqual(a_diff, datetime.timedelta(milliseconds=0),
                            delta=datetime.timedelta(microseconds=500))
 
     # Called in 5 milliseconds

--- a/python/ola/ClientWrapperTest.py
+++ b/python/ola/ClientWrapperTest.py
@@ -173,6 +173,7 @@ class ClientWrapperTest(unittest.TestCase):
     wrapper.AddEvent(datetime.timedelta(milliseconds=5), b)
     wrapper.AddEvent(10, c)
 
+    # Nothing has been called yet
     self.assertIsNone(results.a_called)
     self.assertIsNone(results.b_called)
     self.assertIsNone(results.c_called)
@@ -181,19 +182,32 @@ class ClientWrapperTest(unittest.TestCase):
     self.start = datetime.datetime.now()
     wrapper.Run()
 
+    # Everything has been called
     self.assertIsNotNone(results.a_called)
     self.assertIsNotNone(results.b_called)
     self.assertIsNotNone(results.c_called)
     self.assertIsNotNone(results.d_called)
 
-    self.assertTrue(results.a_called - self.start <
-                    datetime.timedelta(milliseconds=5))
-    self.assertTrue(results.b_called - self.start >=
-                    datetime.timedelta(milliseconds=5))
-    self.assertTrue(results.c_called - self.start >=
-                    datetime.timedelta(milliseconds=10))
-    self.assertTrue(results.d_called - self.start >=
-                    datetime.timedelta(milliseconds=15))
+    # Check when the callbacks were called. Allow 500 microseconds of drift.
+    # Called immediately
+    a_diff = results.a_called - self.start
+    self.assertAlmostEqual(a_diff, datetime.timedelta(microseconds=0),
+                           delta=datetime.timedelta(microseconds=500))
+
+    # Called in 5 milliseconds
+    b_diff = results.b_called - self.start
+    self.assertAlmostEqual(b_diff, datetime.timedelta(milliseconds=5),
+                           delta=datetime.timedelta(microseconds=500))
+
+    # Called in 10 milliseconds
+    c_diff = results.c_called - self.start
+    self.assertAlmostEqual(c_diff, datetime.timedelta(milliseconds=10),
+                           delta=datetime.timedelta(microseconds=500))
+
+    # Called in 15 milliseconds
+    d_diff = results.d_called - self.start
+    self.assertAlmostEqual(d_diff, datetime.timedelta(milliseconds=15),
+                           delta=datetime.timedelta(microseconds=500))
 
     sockets[0].close()
     sockets[1].close()
@@ -229,6 +243,9 @@ class ClientWrapperTest(unittest.TestCase):
     wrapper.Run()
 
     self.assertTrue(results.gotdata)
+
+    sockets[0].close()
+    sockets[1].close()
 
 
 if __name__ == '__main__':

